### PR TITLE
[#13096] Prevent of throwing error with reading path of empty selection.

### PR DESCRIPTION
**Pull request by msamsel**
- Detection when range of selection is empty and prevent of calculating path in such cases.

### CODE REVIEW

File: tests/core/editable/backspaceOrDeleteError.js [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+};
+
+// #13096
+bender.test( {
+    'test deleting text without selection': function() {
+		this.editor.getSelection().removeAllRanges();
+		this.editor.fire( 'key', {
+			domEvent: {
+				getKey: function() {
+					return 8;
```
Comment (Comandeer) [inline tests/core/editable/backspaceOrDeleteError.js]: Let's add test for delete also, for completness.

**Comment by msamsel** [comment]
- Added <kbd>Delete</kbd> key to Unit test.
- Moved test to keystroke.js
- Added manual test
- Moved checking selection outside `if` statement, so it should be more generic.

About <kbd>Enter</kbd>, this is another error, which was should be resolve in this PR @ #359

File: tests/core/editable/keystrokes/keystrokes.js [hunk 2]
```diff
@@ -166,5 +166,34 @@ bender.test( {
...
+		editor.getSelection().removeAllRanges();
+		editor.fire( 'key', {
+			domEvent: {
+				getKey: function() {
+					return DEL;
+				}
+			}
+		} );
+
+		bot.setHtmlWithSelection('<p>^Foo</p>');
```
Comment (Comandeer) [inline tests/core/editable/keystrokes/keystrokes.js]: There are 2 different cases – it'd divide them.

File: tests/core/editable/keystrokes/delbackspacequirks/manual/noselection.md [hunk 3]
```diff
@@ -0,0 +1,14 @@
+@bender-tags: 4.7.0, tc, 13096
+@bender-ui: collapsed
+@bender-ckeditor-plugins: toolbar, wysiwygarea
+
+1. Open browser console.
```
Comment (Comandeer) [inline tests/core/editable/keystrokes/delbackspacequirks/manual/noselection.md]: Actually I'm not able to create situation, in which there will be really no-selection in the editor.

Comment (msamsel) [inline tests/core/editable/keystrokes/delbackspacequirks/manual/noselection.md]: Hmmm.
I see that currently problem might be reproduced only on Chrome browser. FF, IE and Safari seems to don't have such glitch and caret remain in editor for the whole time.
I precise description to use Chrome browser for testing.

Does it help in this case or should I make some more detail description?

Comment (Comandeer) [inline tests/core/editable/keystrokes/delbackspacequirks/manual/noselection.md]: Hm, restricting manual test to only one browser doesn't seem like good solution. Maybe trying to somehow change the test would be better? E.g. add button that removes selection from the editor?

File: tests/core/editable/keystrokes/delbackspacequirks/manual/noselection.md [hunk 4]
```diff
@@ -0,0 +1,15 @@
+@bender-tags: 4.7.0, tc, 13096
```
Comment (Comandeer) [inline tests/core/editable/keystrokes/delbackspacequirks/manual/noselection.md]: Now we could update test tag ;)

File: tests/core/editable/keystrokes/delbackspacequirks/manual/noselection.md [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+@bender-ui: collapsed
+@bender-ckeditor-plugins: toolbar, wysiwygarea
+
+1. Open Chrome browser.
```
Comment (Comandeer) [inline tests/core/editable/keystrokes/delbackspacequirks/manual/noselection.md]: As I stated in the comment, restricting this test to one browser is not the best option. The test should be changed to use more generic method.

File: core/editable.js [hunk 6]
```diff
@@ -998,17 +998,23 @@
 					var keyCode = evt.data.domEvent.getKey(),
 						isHandled;
 
+					// #13096
```
Comment (Comandeer) [inline core/editable.js]: Maybe two words of explanation why this change is needed.

File: core/editable.js [hunk 7]
```diff
@@ -1180,8 +1186,14 @@
 						if ( !( key in backspaceOrDelete ) )
 							return;
 
+						// #13096
```
Comment (Comandeer) [inline core/editable.js]: Explanation could be nice also here.

**Comment by mlewand** [comment]
@msamsel @Comandeer One thing that have not been mentioned here is that removing `.DS_Store` is outside of the scope of this ticket and should be done in a separate contribution.

File: tests/core/editable/keystrokes/delbackspacequirks/manual/noselection.md (hunk 5 above)
Comment (msamsel) [inline tests/core/editable/keystrokes/delbackspacequirks/manual/noselection.md]: @Comandeer can you explain how do you see that? I can restrict entire test to one browser but I'm not sure that this is the case here.
Noting else came to my mind about improving this test to be more generic. I'd appreciate a little help or some idea which might help me enchance this part. :)

Comment (msamsel) [inline tests/core/editable/keystrokes/delbackspacequirks/manual/noselection.md]: Ohh I see your comment above. Still getting used to github ;)
I'll make change after weekend

**Comment by msamsel** [comment]
Comments and version was updated.
.DS_Store was restored and removing will be requested in another PR.

After weekend I made changes to manual test

File: tests/core/editable/keystrokes/delbackspacequirks/manual/noselection.md (hunk 4 above)
Comment (msamsel) [inline tests/core/editable/keystrokes/delbackspacequirks/manual/noselection.md]: done

File: core/editable.js (hunk 6 above)
Comment (msamsel) [inline core/editable.js]: done

File: core/editable.js (hunk 7 above)
Comment (msamsel) [inline core/editable.js]: and done

File: tests/core/editable/keystrokes/delbackspacequirks/manual/noselection.html [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+	}
+
+	CKEDITOR.replace( 'editor', {
+		height: 300
+	} );
+</script>
+
+<button onclick="start()" >Start.</button>
+<script>
+	function start() {
```
Comment (Comandeer) [inline tests/core/editable/keystrokes/delbackspacequirks/manual/noselection.html]: Why so complicated? 😱 

```javascript
function start() {
    var ed = CKEDITOR.instances.editor;
    ed.focus();
    ed.getSelection().removeAllRanges();
}
```
seems to do the trick on Chrome and Firefox, so I suppose that it will also work in other browsers.

File: tests/core/editable/keystrokes/delbackspacequirks/manual/noselection.md [hunk 9]
```diff
@@ -0,0 +1,19 @@
+@bender-tags: 4.7.1, tc, 13096
+@bender-ui: collapsed
+@bender-ckeditor-plugins: toolbar, wysiwygarea
+
+1. Open browser console.
```
Comment (Comandeer) [inline tests/core/editable/keystrokes/delbackspacequirks/manual/noselection.md]: If you simplify the above function, then most of these steps could be deleted.

File: tests/core/editable/keystrokes/delbackspacequirks/manual/noselection.html (hunk 8 above)
Comment (msamsel) [inline tests/core/editable/keystrokes/delbackspacequirks/manual/noselection.html]: `Button` get focus after pressing it.
What more I cannot just use `ed.focus()` because in such case selection automatically shows caret at the beginning of line.
I cannot also click into editor to restore focus in editor, because such scenario also show caret.
Idea is to remove ranges without loosing focus on editor.

What just came on my mind is to create some div 20px x 20px, which might simulate similar case with `mouseenter` event. Because `click` event mess up with focus I need to avoid them in this case. What do you think about that?

Comment (msamsel) [inline tests/core/editable/keystrokes/delbackspacequirks/manual/noselection.html]: Ok I test it, and it's working as it should. So it simplifies entire test :) I make a new commit with such improvement.

File: tests/core/editable/keystrokes/delbackspacequirks/manual/noselection.md (hunk 9 above)
Comment (msamsel) [inline tests/core/editable/keystrokes/delbackspacequirks/manual/noselection.md]: After change description is now shorter and simpler.

**Comment by Comandeer** [comment]
Added issue for this PR: https://github.com/ckeditor/ckeditor-dev/issues/457
